query-engine/js-connectors/js/neon-js-connector/src/neon.ts

```diff
@@ -18,22 +18,31 @@ const TRANSACTION_ROLLBACK = 'ROLLBACK'
 class PrismaNeon implements Connector, Closeable {
   readonly flavor = 'postgres'
 
-  private client: Client
+  private client?: Client
+  private httpClient?: NeonQueryFunction<false, true>
   private isRunning: boolean = true
   private inTransaction: boolean = false
   private _isHealthy: boolean = true
   private _version: string | undefined = undefined
+  private _httpMode: boolean
 
   constructor(config: PrismaNeonConfig) {
-    const { url: connectionString, ...rest } = config
-    this.client = new Client({ connectionString, ...rest })
-    // connect the client in the background, all requests will be queued until connection established
-    this.client.connect()
+    const { url: connectionString, httpMode, ...rest } = config
+    this._httpMode = httpMode ?? false
+    if (!this._httpMode) {
+      this.client = new Client({ connectionString, ...rest })
+      // connect the client in the background, all requests will be queued until connection established
+      this.client.connect()
+    } else {
+      this.httpClient = neon(connectionString, { fullResults: true, ...rest })
+    }
```

Comment:
We could take inspiration from [this](https://github.com/prisma/prisma-engines/blob/3b74ba351abfefd10f3da2b8af3c7ed7658f8d30/query-engine/js-connectors/js/planetscale-js-connector/src/planetscale.ts#L13-L35) and store the driver's mode and corresponding client within the same object, as to leverage TypeScript rather than working around it.

Given something like:

```typescript
type ModeSpecificDriver
  = {
    /**
     * Indicates that we're using the HTTP mode.
     */
    mode: 'http'

    /**
     * The Neon HTTP client, without transaction support.
     */
    client: NeonQueryFunction<ARRAY_MODE, FULL_RESULTS>
  }
  | {
    /**
     * Indicates that we're using the WebSocket mode.
     */
    mode: 'ws'

    /**
     * The standard Neon client, with transaction support.
     */
    client: Client
  }
```

The `constructor` could become something like:

```typescript
  constructor(config: PrismaNeonConfig) {
    const { url: connectionString, mode, ...rest } = config

    if (mode === 'http') {
      this.driver = {
        mode: 'http',
        client: neon(connectionString, { fullResults: true, ...rest }),
      }
    } else {
      this.driver = {
        mode: 'ws',
        client: new Client({ connectionString, ...rest }),
      }

      // connect the client in the background, all requests will be queued until connection established
      this.driver.client.connect()
    }
  }
```

Reply:
looks good! I was looking for something like Rust enum type in TypeScript and this looks really well.